Implement `@dynamicCallable`.

dan-zheng opened the pull request:
This PR introduces the `@dynamicCallable` attribute, which enables nominal types to be "callable" via a simple syntactic sugar.

This is the implementation of [SE-0216](https://github.com/apple/swift-evolution/blob/master/proposals/0216-dynamic-callable.md). Read the proposal for more details.

---

`@dynamicCallable` is a follow-up to [SE-0195 "Dynamic Member Lookup"](https://github.com/apple/swift-evolution/blob/master/proposals/0195-dynamic-member-lookup.md).

Forum discussion regarding `@dynamicCallable`:
  - [Pitch #1: Introduce user-defined dynamically "callable" types](https://forums.swift.org/t/pitch-introduce-user-defined-dynamically-callable-types/7038)
  - [Pitch #2: Introduce user-defined dynamically “callable” types](https://forums.swift.org/t/pitch-2-introduce-user-defined-dynamically-callable-types/7112)
  - [Pitch #3: Introduce user-defined dynamically “callable” types](https://forums.swift.org/t/pitch-3-introduce-user-defined-dynamically-callable-types/12232)

Proposal review and acceptance rationale:
  - [Proposal review: SE-216: User-defined dynamically callable types](https://forums.swift.org/t/se-0216-user-defined-dynamically-callable-types/13615)
  - [[Accepted] SE-216: User-defined dynamically callable types](https://forums.swift.org/t/accepted-se-216-user-defined-dynamically-callable-types/14110)

---

Implementation notes:

* The `@dynamicCallable` implementation has been revamped according to [proposal clarifications](https://github.com/apple/swift-evolution/pull/948).
  * The main clarification is about the precise rules for dynamic call ambiguity resolution.
  * The rules are based solely on call-site syntax.
    * If a `@dynamicCallable` type implements the `withArguments:` method and it is called with no keyword arguments, use the `withArguments:` method.
    * In all other cases, attempt to use the `withKeywordArguments:` method.
  * The rules are intentionally simple. Advanced return type overloading ([example](https://github.com/apple/swift/pull/16980#discussion_r204202532)) is not supported.
* `@dynamicCallable` types with multiple `dynamicallyCall` methods are supported.
  * A type may have multiple `dynamicallyCall` methods if it inherits from a `@dynamicCallable` superclass or conforms to `@dynamicCallable` protocols, or if it itself defines overloaded `dynamicallyCall` methods.
  * To support multiple `dynamicallyCall` methods, the implementation creates a disjunction constraint over all `dynamicallyCall` methods, and pairs it with a `DynamicCallableApplicableFunction` constraint.
    * Originally, I tried pairing the disjunction with an `ApplicableFunction` constraint, but encountered [a crash involving a parameter label count mismatch in `constraints::matchCallArguments`](https://github.com/apple/swift/pull/16980#issuecomment-423710872).
      * This crash occurs because dynamic calls are not an exact sugar: the dynamic call expression may contain multiple parameters (e.g. three parameters in `x(1, 2, 3)`) but the `dynamicallyCall` method defines only one parameter (e.g. one parameter in `x.dynamicallyCall(withArguments: [Int]`).
      * AFAIK, there's no precedent for such mismatches (where the parameter count for a `CallExpr` and the parameter count of the function it calls don't match).
    * To work around the crash, I found it necessary to add a new `DynamicallyCallableApplicableFunction` constraint kind. `simplifyDynamicCallableApplicableFnConstraint` has custom logic for checking parameter types and doesn't produce the crash.
* Please look at [`test/attr/attr_dynamic_callable.swift`](https://github.com/apple/swift/pull/20305/files#diff-9fa29220a2e4e5078d3ada720a54e370) to see what functionality is supported/unsupported.

dan-zheng committed:
Implement dynamically callable types (`@dynamicCallable`).

- Implement dynamically callable types as proposed in SE-0216.
  - Dynamic calls are resolved based on call-site syntax.
  - Use the `withArguments:` method if it's defined and there are no
    keyword arguments.
  - Otherwise, use the `withKeywordArguments:` method.
- Support multiple `dynamicallyCall` methods.
  - This enables two scenarios:
    - Overloaded `dynamicallyCall` methods on a single
      `@dynamicCallable` type.
    - Multiple `dynamicallyCall` methods from a `@dynamicCallable`
      superclass or from `@dynamicCallable` protocols.
  - Add `DynamicCallableApplicableFunction` constraint. This, used with
    an overload set, is necessary to support multiple `dynamicallyCall`
    methods.

dan-zheng commented:
The implementation has been revised to match the proposal.
- Dynamic calls are resolved based on call-site syntax.
  - Use the `withArguments:` method if it's defined and there are no keyword arguments.
  - Otherwise, use the `withKeywordArguments:` method.

Additionally, I decided to support multiple `dynamicallyCall` methods.
This makes `@dynamicCallable` dynamic call resolution match the behavior of normal protocol requirement resolution and `@dynamicMemberLookup` resolution.
  - This enables two scenarios:
    - Overloaded `dynamicallyCall` methods on a single `@dynamicCallable` type.
    - Multiple `dynamicallyCall` methods from a `@dynamicCallable` superclass or from `@dynamicCallable` protocols.
  - Add `DynamicCallableApplicableFunction` constraint. This, used with an overload set, is necessary to support multiple `dynamicallyCall` methods.

Please look at the [test file](https://github.com/apple/swift/pull/20305/files#diff-9fa29220a2e4e5078d3ada720a54e370) to see what's supported/unsupported.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please smoke test

lattner commented:
Awesome Dan!

dan-zheng committed:
[NFC] Add/clarify comments.

rudkx commented:
@swift-ci build toolchain

dan-zheng commented:
@rudkx I wonder if you (or other code owners) have the time to please review this PR? 
I'd really like to land it before the Swift 5 cutoff. Thank you for all your feedback/advice so far!

Here are some notes for reviewers:
* The `@dynamicCallable` implementation has been revamped according to [proposal clarifications](https://github.com/apple/swift-evolution/pull/948).
  * The main clarification is about the precise rules for dynamic call ambiguity resolution.
  * The rules are succinctly described as:
    * If a `@dynamicCallable` type implements the `withArguments:` method and it is called with no keyword arguments, use the `withArguments:` method.
    * In all other cases, attempt to use the `withKeywordArguments:` method.
  * The rules are intentionally simple. Advanced return type overloading ([example](https://github.com/apple/swift/pull/16980#discussion_r204202532)) is not supported.
* `@dynamicCallable` types with multiple `dynamicallyCall` methods are supported.
  * A type may have multiple `dynamicallyCall` methods if it inherits from a `@dynamicCallable` superclass or conforms to `@dynamicCallable` protocols, or if it itself defines overloaded `dynamicallyCall` methods.
  * To support multiple `dynamicallyCall` methods, the implementation creates a disjunction constraint over all `dynamicallyCall` methods, and pairs it with a `DynamicCallableApplicableFunction` constraint.
    * Originally, I tried pairing the disjunction with an `ApplicableFunction` constraint, but encountered a [crash involving a parameter label count mismatch in `constraints::matchCallArguments`](https://github.com/apple/swift/pull/16980#issuecomment-423710872).
      * This crash occurs because dynamic calls are not an exact sugar: the dynamic call expression contains multiple parameters (e.g. three parameters in `x(1, 2, 3)`) but the `dynamicallyCall` method defines only one parameter (e.g. one parameter in `x.dynamicallyCall(withArguments: [Int]`).
      * AFAIK, there's no precedent for such mismatches (where the parameter count for a `CallExpr` and the parameter count of the function it calls don't match).
    * To work around the crash, I found it necessary to add a new `DynamicallyCallableApplicableFunction` constraint kind. `simplifyDynamicCallableApplicableFnConstraint` has custom logic for checking parameter types and doesn't produce the crash.
* Please look at [`test/attr/attr_dynamic_callable.swift`](https://github.com/apple/swift/pull/20305/files#diff-9fa29220a2e4e5078d3ada720a54e370) to see what functionality is supported/unsupported.

dan-zheng commented:
I added some implementation notes to the PR description.

@rudkx I wonder if you (or other code owners) have the time to please review this PR? 
I'd really like to land it before the Swift 5 cutoff. Thank you for all your feedback/advice so far!

lattner commented on lib/Sema/CSApply.cpp:
```diff
@@ -7366,67 +7434,72 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+                                   /*dotLoc=*/SourceLoc(), choice,
+                                   DeclNameLoc(fn->getEndLoc()),
+                                   selected->openedType, locator, ctorLocator,
+                                   /*Implicit=*/true, choice.getFunctionRefKind(),
```
Please fit in 80 columns.

lattner commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4497,7 +4504,280 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+    auto cand = cast<FuncDecl>(choice.getDecl());
+    return !isValidDynamicCallableMethod(cand, decl, CS.TC, hasKeywordArgs);
+  };
+  candidates.erase(std::remove_if(candidates.begin(), candidates.end(), filter),
```
Here too :)

dan-zheng replied:
Very minor, but:

This line was actually at 80 columns long, actually (i.e. the last character was on column 80).
I wonder if the 80 column convention is:
* The last character of a line can be on column 80 (or before). I've been following this.
* The last character of a line can be on column 79 (or before). I suppose this makes the line "fit within 80 columns".

Anyhow, I changed this line and other lines that end on column 80 so that they end on column 79.

lattner commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4497,7 +4504,280 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+    lookupDynamicCallableMethods(type, CS, locator, ctx.Id_withArguments,
+                                 /*hasKeywordArgs*/ false);
+  methods.keywordArgumentsMethods =
+    lookupDynamicCallableMethods(type, CS, locator, ctx.Id_withKeywordArguments,
```
Here too

lattner reviewed (approved):
Please make sure the code fits in 80 columns, but it otherwise looks ok to me.  

@rudkx do you mind if Dan merges this?  We can continue to improve it in master.

dan-zheng committed:
[NFC] Minor edits.

- Fit lines within 80 columns (addresses feedback by @lattner).
- Rename `DictionaryLiteral` to `KeyValuePairs`.

dan-zheng commented:
Thank you @lattner and @rudkx!
@lattner: I addressed in your comments in https://github.com/apple/swift/pull/20305/commits/c4189f3a3a6aabd7591c01d399e93c5ea4bd87ae.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

dan-zheng commented:
It seems the source compatibility suite failed due to a CI error: [`FATAL: command execution failed`.](https://ci.swift.org/job/swift-PR-source-compat-suite/2133/console)
Rerunning. @swift-ci Please test source compatibility

dan-zheng committed:
[NFC] Minor edits.

Fix indentation and tweak comments.

dan-zheng commented:
All smoke/source compatibility tests passed!
I made some final NFC edits in https://github.com/apple/swift/pull/20305/commits/14e28ac92e2649e9ad7eec2748c56179bedd361e. Retriggering tests now.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please test source compatibility

dan-zheng commented:
To reviewers: I'll leave merging the PR to you. After you've reviewed and feel confident.

lattner commented:
Thank you Dan!!!  Please update the ChangeLog.md file if you haven't already!

dan-zheng commented:
Yay! I'm happy to land this PR after first opening it in June. 😛Thanks for your patience!
I'd like to express thanks for everyone who gave me guidance and feedback along the way.

I'll update the changelog soon.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4497,7 +4504,283 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+  // If the right-hand side is not a function type, it must be a valid
+  // @dynamicCallable type. Attempt to get valid `dynamicallyCall` methods.
+  auto methods = getDynamicCallableMethods(desugar2, *this, locator);
+  if (!methods.isValid()) return SolutionKind::Error;
```
@slavapestov: [you gave this feedback on the old implementation PR](https://github.com/apple/swift/pull/16980#issuecomment-418849481):
> Also you should not emit diagnostics during constraint simplification because we can explore different parts of the solution space as we visit disjunctions, and discard invalid paths.

I made sure not to emit diagnostics here, during simplification. Instead, I return `SolutionKind::Error` and added diagnostic logic in `CSDiag.cpp`.

dan-zheng commented:
FYI: I'll try to push two extra PRs before the Swift 5 cutoff:
- [ ] Fix `@dynamicMemberLookup` protocols.
  * Currently, `subscript(dynamicMember:)` does not work as a `@dynamicMemberLookup` protocol requirement.
  * I fixed this for `@dynamicCallable`, need to backport the logic to `@dynamicMemberLookup`.
- [ ] Polish `@dynamicMemberLookup` implementation so that it matches the style of the `@dynamicCallable` implementation.
  * This involves making consistent things like variable names, doc comment format, etc.
